### ISSUE TITLE
File::Find documentation - is "Don't modify these variables" still valid?

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8416]
Migrated from [rt.perl.org#38965](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=38965) (status was 'resolved')

Searchable as RT38965$

**Comment by p5pRT** [comment]
### From adavies@ptc.com

#### Created by adavies@ptc.com



In the doucmentation for File&#8203;::Find\, it says of the variables
C\<$File&#8203;::Find&#8203;::dir>\, C\<$File&#8203;::Find&#8203;::name> and C\<$\_> not to
modify these variables\.

Is this still true\, especially in the case of C\<$\_>?
Looking at Find\.pm\, it looks safe to modify them in the callback
without affecting File&#8203;::Find's functionality\.

The reason i ask\, is that in the case where you want to
use File&#8203;:Find&#8203;::find to read thro' files\, using something like&#8203;:
C\<open F\, $\_ or die; while \(\<F>\) \{ \.\.\. \}>\, the above
documentation suggests you need to start your callback
with C\<local $\_ = $\_;> to ensure you "don't modify" C\<$\_>\.
Unfortunately\, this seemingly innocuous addition can give
a major performance slowdown&#8203;:

\# %\<

use warnings;
use strict;
use File&#8203;::Find;
use Benchmark;

my $CODE = \<\<'EOT';
&nbsp;    open\(F\, "\<"\, $\_\) or return;
&nbsp;    while \(\<F>\) \{\}
&nbsp;    close F;

EOT

eval \<\<"EOT";
&nbsp;    sub search\_with\_local\_dollar\_underscore \{
&nbsp;        local \\$\_ = \\$\_;
&nbsp;        $CODE
&nbsp;    \}
EOT

eval \<\<"EOT";
&nbsp;    sub search\_WITHOUT\_local\_dollar\_underscore \{
&nbsp;        $CODE
&nbsp;    \}
EOT

my $dir = 'demedir';
\-d $dir or mkdir $dir or die "cannot mkdir $dir&#8203;: $\!\\n";
if \(1\) \{
&nbsp;    my $file = "$dir/bigline";
&nbsp;    print "Creating $file\\n";
&nbsp;    open FOUT\, ">&#8203;:raw"\, $file or die "cannot open $file&#8203;: $\!\\n";
&nbsp;    print FOUT "?" x 5000000;
&nbsp;    close FOUT;
\}

my $N = 2;
Benchmark&#8203;::timethese\($N\, \{
&nbsp;    'search\_with\_local\_dollar\_underscore' => sub \{
&nbsp;        File&#8203;::Find&#8203;::find\(\\&search\_with\_local\_dollar\_underscore\,
$dir\);
&nbsp;    \}\,
&nbsp;    'search\_WITHOUT\_local\_dollar\_underscore' => sub \{
&nbsp;    
File&#8203;::Find&#8203;::find\(\\&search\_WITHOUT\_local\_dollar\_underscore\, $dir\);
&nbsp;    \}\,
\}\);

\# END \{ rmdir $dir or warn "cannot rmdir $dir&#8203;: $\!\\n" \}

\# >%

I get&#8203;:

 Creating demedir/bigline
 Benchmark&#8203;: timing 2 iterations of
search\_WITHOUT\_local\_dollar\_underscore\,
search\_with\_local\_dollar\_underscore\.\.\.
 search\_WITHOUT\_local\_dollar\_underscore&#8203;:
&nbsp;  6 wallclock secs \( 2\.08 usr \+  3\.66 sys =  5\.73 CPU\) @&#8203;  0\.35/s \(n=2\)
 search\_with\_local\_dollar\_underscore&#8203;:
 12 wallclock secs \( 4\.47 usr \+  7\.09 sys = 11\.56 CPU\) @&#8203;  0\.17/s \(n=2\)


I came across this\, when trying to understand why a search
program was running slower and tracked it down to where it
was processing large bitmaps files\. Removing the initial
C\<local $\_ = $\_;> from the search routine sped things up
dramatically\.

Thanks\.

<details><summary>Perl Info</summary>

```

Flags:
    category=docs
    severity=low

Site configuration information for perl v5.8.7:

Configured by Andy_Cuss at Thu Aug 11 14:02:10 2005.

Summary of my perl5 (revision 5 version 8 subversion 7) configuration:
  Platform:
    osname=MSWin32, osvers=5.1, archname=MSWin32-x86-multi-thread
    uname=''
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    usethreads=define use5005threads=undef useithreads=define
usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cl', ccflags ='-nologo -Gf -W3 -MD -DNDEBUG -O1 -DWIN32
-D_CONSOLE -DNO_STRICT -DHAVE_DES_FCRYPT  -DPERL_IMPLICIT_CONTEXT
-DPERL_IMPLICIT_SYS -DUSE_PERLIO -DPERL_MSVCRT_READFIX',
    optimize='-MD -DNDEBUG -O1',
    cppflags='-DWIN32'
    ccversion='12.00.8804', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='__int64',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -release
-libpath:"c:\perl3\lib\CORE"  -machine:x86'
    libpth=\lib
    libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib
comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib
netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib
odbc32.lib odbccp32.lib msvcrt.lib
    perllibs=  oldnames.lib kernel32.lib user32.lib gdi32.lib
winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib
oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib
version.lib odbc32.lib odbccp32.lib msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=yes, libperl=perl58.lib
    gnulibc_version='undef'
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -release
-libpath:"c:\perl3\lib\CORE"  -machine:x86'

Locally applied patches:
    


@INC for perl v5.8.7:
    C:/perl3/lib
    C:/perl3/site/lib
    .


Environment for perl v5.8.7:
    HOME=C:\alex
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=C:\Program Files\Tcl-8.5.0.0\bin;C:\perl3\bin;C:\Program
Files\Tcl-8.4.8.0\bin;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32
\Wbem;C:\alex\bin;C:\Program Files\ATI Technologies\ATI Control
Panel;C:\Program Files\Perforce;C:\cygwin\bin;C:\Program Files\Crimson
Editor;C:\Program Files\Microsoft Visual Studio\VC98\Bin;C:\Program
Files\Microsoft Visual Studio\Common\Tools;C:\Program
Files\Vantive32;C:\Program Files\QuickTime\QTSystem\;C:\Program
Files\cvsnt;C:\Program Files\Sun\AppServer\bin;C:\ghc\ghc-6.2.2\bin
    PERL_BADLANG (unset)
    SHELL (unset)

```
</details>

**Comment by p5pRT** [comment]
### From dland@landgren.net

> \[adavies@&#8203;ptc\.com \- Sun Apr 23 07&#8203;:37&#8203;:19 2006\]&#8203;:
> 
> This is a bug report for perl from adavies@&#8203;ptc\.com\,
> generated with the help of perlbug 1\.35 running under perl v5\.8\.7\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> In the doucmentation for File&#8203;::Find\, it says of the variables
> C\<$File&#8203;::Find&#8203;::dir>\, C\<$File&#8203;::Find&#8203;::name> and C\<$\_> not to
> modify these variables\.
> 
> Is this still true\, especially in the case of C\<$\_>?
> Looking at Find\.pm\, it looks safe to modify them in the callback
> without affecting File&#8203;::Find's functionality\.
> 
> The reason i ask\, is that in the case where you want to
> use File&#8203;:Find&#8203;::find to read thro' files\, using something like&#8203;:
> C\<open F\, $\_ or die; while \(\<F>\) \{ \.\.\. \}>\, the above
> documentation suggests you need to start your callback
> with C\<local $\_ = $\_;> to ensure you "don't modify" C\<$\_>\.
> Unfortunately\, this seemingly innocuous addition can give
> a major performance slowdown&#8203;:

The restriction still holds\. You don't notice it because the benchmark
finds only one file in one directory\.

That said\, the documentation could certainly provide better advice\.
Localising $\_ \(to nothing in particular\) after it has been used to fetch
the file name results in much better performance\.

sub search\_with\_local\_dollar\_underscore\_smart \{
&nbsp;    open\(F\, "\<"\, $\_\) or return;
&nbsp;    local $\_; \# do it now\!
&nbsp;    while \(\<F>\) \{\}
&nbsp;    close F;
\}javascript&#8203;:toggleCc\(\)
Click to Toggle Cc of perl5\-porters@&#8203;perl\.org

I'm not sure that the benchmark proves much\. You should create many
files\, rather than just one\, otherwise you won't know if $\_ breaks \(and
modifying your code shows that it does\)\.

But it shows that there is something weird going on in Win32 land\.
Running your code on Unix requires upping the iterations to around 10000
to get a meaningful difference but I see the same behaviour on Win32\. I
tried taking out the '&#8203;:raw' on the open\, without any significant
difference\. I'm not sure what the reason is\. Changing the file 'bigfile'
to 10000 lines of 500 chars made no difference either\.

HTH\,
David

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From david@landgren.net

Davies\, Alex \(via RT\) wrote&#8203;:
> \# New Ticket Created by  "Davies\, Alex" 
> \# Please include the string&#8203;:  \[perl \#38965\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=38965 >
> 
> 
> This is a bug report for perl from adavies@&#8203;ptc\.com\,
> generated with the help of perlbug 1\.35 running under perl v5\.8\.7\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> In the doucmentation for File&#8203;::Find\, it says of the variables
> C\<$File&#8203;::Find&#8203;::dir>\, C\<$File&#8203;::Find&#8203;::name> and C\<$\_> not to
> modify these variables\.

It is perhaps worth spelling out one of the few remaining legitimate 
reasons to use C\<local>\, since the bug reports shows that someone 
thought C\<local $\_ = $\_> was necessary for it to work\, when C\<local $\_> 
is sufficient\.

David\.
\-\- 
"It's overkill of course\, but you can never have too much overkill\."

**Comment by p5pRT** [comment]
### From david@landgren.net

<details><summary>Find.pm.patch</summary>

```diff
--- Find.pm.orig	Mon Apr 24 12:40:14 2006
+++ Find.pm	Mon Apr 24 12:56:29 2006
@@ -215,8 +215,6 @@
 
 =back
 
-Don't modify these variables.
-
 For example, when examining the file F</some/path/foo.ext> you will have:
 
     $File::Find::dir  = /some/path/
@@ -239,6 +237,18 @@
               /etc               /                 /etc
               /etc/x             /etc              /etc/x
 
+Do not modify these variables. If you want to use C<$_>, it must
+be restored to its initial value before returning from the
+function. C<local> may be used for this purpose:
+
+  sub callback {
+    open my $fh, '<', $_ or die "Cannot open $_ for input: $!\n";
+    local $_; # localize $_ for the remainder of the routine
+    while (<$fh>) {
+      # manipulate $_
+    }
+    # $_ will be restored upon leaving
+  }
 
 When <follow> or <follow_fast> are in effect, there is
 also a C<$File::Find::fullname>.  The function may set

```
</details>

**Comment by p5pRT** [comment]
### From alex_davies@3b2.com

> \[dland \- Mon Apr 24 02&#8203;:57&#8203;:34 2006\]&#8203;:
> 
> > \[adavies@&#8203;ptc\.com \- Sun Apr 23 07&#8203;:37&#8203;:19 2006\]&#8203;:
> > The reason i ask\, is that in the case where you want to
> > use File&#8203;:Find&#8203;::find to read thro' files\, using something like&#8203;:
> > C\<open F\, $\_ or die; while \(\<F>\) \{ \.\.\. \}>\, the above
> > documentation suggests you need to start your callback
> > with C\<local $\_ = $\_;> to ensure you "don't modify" C\<$\_>\.
> > Unfortunately\, this seemingly innocuous addition can give
> > a major performance slowdown&#8203;:
> 
> The restriction still holds\. You don't notice it because the benchmark
> finds only one file in one directory\.
> 

Sorry \- the testcase was \_only\_ to show the performance impact
of using C\<local $\_> and not that you could or could not modify
the C\<$\_>\, C\<$name> or C\<$dir>\.
The following test shows that is NOT ok to change
C\<$File&#8203;::Find&#8203;::dir>\, but for the other 2 it \_might\_ be&#8203;:

\# %\<
use warnings;
use strict;
use File&#8203;::Find;
$| = 1;
my $dir = '/windows'; \# or whatever\.
my \(@&#8203;files1\, @&#8203;files2\);

print "find1\\n"; \# 
File&#8203;::Find&#8203;::find\(sub \{
&nbsp;    push @&#8203;files1\, \[$File&#8203;::Find&#8203;::name\, $File&#8203;::Find&#8203;::dir\, $\_\];
&nbsp;    $File&#8203;::Find&#8203;::name = \['boo'\];
&nbsp;    \#\#$File&#8203;::Find&#8203;::dir = \{1 => 3\};
&nbsp;    $\_ = \\33;
\}\, $dir\);

print "find2\\n";
File&#8203;::Find&#8203;::find\(sub \{
&nbsp;    push @&#8203;files2\, \[$File&#8203;::Find&#8203;::name\, $File&#8203;::Find&#8203;::dir\, $\_\];
\}\, $dir\);

die "different number of files" unless @&#8203;files1 == @&#8203;files2;

for \(my $i = 0; $i \< @&#8203;files1; \+\+$i\) \{
&nbsp;    my \($n1\, $d1\, $u1\) = @&#8203;\{ $files1\[$i\] \};
&nbsp;    my \($n2\, $d2\, $u2\) = @&#8203;\{ $files2\[$i\] \};
&nbsp;    if \($n1 ne $n2 or \(0 and $d1 ne $d2\) or $u1 ne $u2\) \{
&nbsp;        die "i=$i 1&#8203;:$n1\,$d1\,$u1 2&#8203;:$n2\,$d2\,$u2\\n";
&nbsp;    \}
\}
print "ok\\n";

\# >%

Rereading Find\.pm suggests this is right\.
Although i'm only really concerned about the need to localise C\<$\_>\,
as changing C\<$dir> and C\<$name> is much less common\.

> That said\, the documentation could certainly provide better advice\.
> Localising $\_ \(to nothing in particular\) after it has been used
> to fetch the file name results in much better performance\.
> 
> sub search\_with\_local\_dollar\_underscore\_smart \{
>     open\(F\, "\<"\, $\_\) or return;
>     local $\_; \# do it now\!
>     while \(\<F>\) \{\}
>     close F;
> \}

Even with this\, i still see a performance hit\.

> But it shows that there is something weird going on in Win32 land\.
> Running your code on Unix requires upping the iterations to around
> 10000 to get a meaningful difference but I see the same behaviour
> on Win32\. I tried taking out the '&#8203;:raw' on the open\, without any
> significant difference\.

The :raw was a vestige of my first testcase that wrote 0xff's
\(a la a bitmap file\)\, but then realised that any character gave
the same behaviour\)\.

> I'm not sure what the reason is\. Changing
> the file 'bigfile' to 10000 lines of 500 chars made no difference
> either\.
> 
> HTH\,
> David
> 

Thanks\.

**Comment by p5pRT** [comment]
### From @smpeters

> \[david@&#8203;landgren\.net \- Mon Apr 24 04&#8203;:02&#8203;:53 2006\]&#8203;:
> 
> Davies\, Alex \(via RT\) wrote&#8203;:
> > \# New Ticket Created by  "Davies\, Alex" 
> > \# Please include the string&#8203;:  \[perl \#38965\]
> > \# in the subject line of all future correspondence about this issue\. 
> > \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=38965 >
> > 
> > 
> > This is a bug report for perl from adavies@&#8203;ptc\.com\,
> > generated with the help of perlbug 1\.35 running under perl v5\.8\.7\.
> > 
> > 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please enter your report here\]
> > 
> > In the doucmentation for File&#8203;::Find\, it says of the variables
> > C\<$File&#8203;::Find&#8203;::dir>\, C\<$File&#8203;::Find&#8203;::name> and C\<$\_> not to
> > modify these variables\.
> 
> It is perhaps worth spelling out one of the few remaining legitimate 
> reasons to use C\<local>\, since the bug reports shows that someone 
> thought C\<local $\_ = $\_> was necessary for it to work\, when C\<local $\_> 
> is sufficient\.
> 

I applied this patch and then reverted it after chatting with Rafael out
of channel\.  In the current code\, $\_ was localized since File&#8203;::Find 1\.04
\(change \#18501 for those keeping score at home\)\.  I instead fixed the
documenation to reflect this with change \#27955\.

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From adavies@ptc.com

> \-\-\-\-\-Original Message\-\-\-\-\-
> From&#8203;: Steve Peters via RT \[mailto&#8203;:perlbug\-followup@&#8203;perl\.org\] 
> Sent&#8203;: 24 April 2006 17&#8203;:46
> To&#8203;: Davies\, Alex
> Subject&#8203;: \[perl \#38965\] \[RESOLVED\] File&#8203;::Find documentation \- is
> "Don't modify these variables" still valid? 
> 
> According to our records\, your request regarding 
>   "File&#8203;::Find documentation \- is "Don't modify these variables" still
> valid?"  has been resolved\. 
> 
> If you have any further questions or concerns\, please respond to this
> message\.
> 

Looking at the patch \- it now suggests that C\<$File&#8203;::Find&#8203;::dir>
can be modified\.

I don't think this is the case\. Consider line ~883&#8203;:

for my $FN \(@&#8203;filenames\) \{
&nbsp;  next if $FN =~ $File&#8203;::Find&#8203;::skip\_pattern;
&nbsp;  $name = $dir\_pref \. $FN; \# $File&#8203;::Find&#8203;::name
&nbsp;  $\_ = \($no\_chdir ? $name : $FN\); \# $\_
&nbsp;  \{ $wanted\_callback\->\(\) \}; \# protect against wild "next"
\}

Changing C\<$dir> from the wanted\_callback does break functionality\,
as subsequent times thro' the above loop will see the modified
value\. \(Also see the code i posted in my follow up message\.\)

My original bug submission was wrong to suggest that all
three variables are modifiable \- apologies there\.
I believe it is safe to modify C\<$File&#8203;::Find&#8203;::name> and
more usefully C\<$\_>\.

Thanks & Best regards\.

> For other topics\, please create a new ticket\.
> 
> Please don't feel obligated to say "Thanks" or "Kudos" or "I owe you
> a beer" \-\- if you respond to this message it will reopen the ticket\.
> If you must\, please send email directly to the person who handled
> your ticket\, and not to the tracking system\.
> 
> \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=38965 >
